src/util/time.js

```diff
@@ -119,32 +112,58 @@ export const findSegmentForPlayerTime = (time, playlist) => {
   // - there will always be a segment.duration
   // - we can start from zero
   // - segments are in time order
-  // - segment.start and segment.end only come
-  //    from syncController
 
-  return findSegmentForTime(time, 'player', playlist);
-};
+  if (!playlist || !playlist.segments || playlist.segments.length === 0) {
+    return null;
+  }
 
-/**
- * Finds a segment that contains the stream time give as an ISO-8601 string.
- *
- * @param {String} streamTime The ISO-8601 streamTime to find a match for
- * @param {Object} playlist A playlist object to search within
- */
-export const findSegmentForStreamTime = (streamTime, playlist) => {
-  let dateTimeObject;
+  let segmentEnd = 0;
+  let segment;
 
-  try {
-    dateTimeObject = new Date(streamTime);
-  } catch (e) {
-    // TODO something here?
+  for (let i = 0; i < playlist.segments.length; i++) {
+    segment = playlist.segments[i];
+
+    // videoTimingInfo is set after the segment is downloaded and transmuxed, and
+    // should contain the most accurate values we have for the segment's player times.
+    //
+    // Use the accurate transmuxedPresentationEnd value if it is available, otherwise fall
+    // back to an estimate based on the manifest derived (inaccurate) segment.duration, to
+    // calculate an end value.
+    segmentEnd = segment.videoTimingInfo ?
+      segment.videoTimingInfo.transmuxedPresentationEnd : segmentEnd + segment.duration;
+
+    if (time <= segmentEnd) {
+      break;
+    }
+  }
+
+  const lastSegment = playlist.segments[playlist.segments.length - 1];
+
+  if (lastSegment.videoTimingInfo &&
+      lastSegment.videoTimingInfo.transmuxedPresentationEnd < time) {
+    // The time requested is beyond the stream end.
     return null;
   }
 
-  // Assumptions:
-  //  - verifyProgramDateTimeTags has already been run
-  //  - live streams have been started
-  return findSegmentForTime(dateTimeObject, 'stream', playlist);
+  if (time > segmentEnd) {
```

Comment:
might be a little clearer to follow if this has a comment saying it's the last segment